File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root==nullptr) return ans;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int s=q.size();
            vector<int> vec;
            for(int i=0;i<s;i++){
                TreeNode* t=q.front();
                q.pop();
                if(t->left!=nullptr) q.push(t->left);
                if(t->right!=nullptr) q.push(t->right);
                vec.push_back(t->val);
            }
            ans.push_back(vec);
        }
        int c=1;
        for(int i=0;i<ans.size();i++){
            vector<int> fd=ans[i];
            if(c==0){
                reverse(fd.begin(),fd.end());
                c=1;
            }else{
                c=0;
            }
            ans[i]=fd;
        }
        return ans;
    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans=new ArrayList<>();
        if(root==null) return ans;
        Queue<TreeNode> q=new  LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int s=q.size();
            List<Integer> vec=new ArrayList<>();
            for(int i=0;i<s;i++){
                TreeNode t=q.peek();
                q.remove();
                if(t.left!=null) q.add(t.left);
                if(t.right!=null) q.add(t.right);
                vec.add(t.val);
            }
            ans.add(vec);
        }
        int c=1;
        for(int i=0;i<ans.size();i++){
            List<Integer> vec=ans.get(i);
            if(c==0){
                Collections.reverse(vec);
                c=1;
            }else{
                c=0;
            }
            ans.set(i, vec);

        }
        return ans;
    }
}
